### DISCUSSION TITLE
How to test file in memory with ruff?

### DISCUSSION THREAD

**Discussion by qarmin** [astral-sh/ruff#3444] [Q&A] [answered]
Recently I wanted to test in my app file python loaded into memory, but I could not find any way how to do this.

Is this possible to achive now with ruff?

**Comment by charliermarsh** [comment, marked as answer]
It's not possible right now. Are you referring to something like a Ruff API that's exposed on the Python side?

**Reply by qarmin** [reply]
I'm talking about any possibility to scan a file from memory from within Python, another Rust project or even as a direct ruff module with access to private functions(in fact for my applications the last point would probably be the most appropriate)

I saw https://github.com/charliermarsh/ruff/blob/main/crates/ruff_cli/src/commands/run.rs, but I don't really know the code, so I cannot figure if this is proper place to search/hack or not.

**Reply by charliermarsh** [reply]
We used to have a public Rust API but we removed it as a lot of the abstractions were still in flux. So the current API isn't very user-friendly for this purpose. The place to look would probably be https://github.com/charliermarsh/ruff/blob/main/crates/ruff_cli/src/diagnostics.rs#L55.
